### PR TITLE
feat: Support connecting to Presto with a PKCS#12 certificate

### PR DESCRIPTION
**Pull request by inglesp**
Docs at https://github.com/ebmdatalab/datalab-team-manual/pull/9.

### CODE REVIEW

**Comment by inglesp** [comment]
No idea why tests are failing -- they pass locally.

44 (out of 49) tests in `test_emis_background.py` are failing with

```
PrestoQueryError(type=INTERNAL_ERROR, name=GENERIC_INTERNAL_ERROR, message="Expected zero to one elements, but found multiple", query_id=20201231_114338_00191_xxx)
```

The traceback is:

```
cohortextractor/study_definition.py:96: in to_dicts
    return self.backend.to_dicts()
cohortextractor/emis_backend.py:60: in to_dicts
    result = self.execute_query()
cohortextractor/emis_backend.py:196: in execute_query
    cursor.execute(sql)
cohortextractor/presto_utils.py:158: in execute
    self._rows = self.cursor.fetchmany()
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/prestodb/dbapi.py:280: in fetchmany
    row = self.fetchone()
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/prestodb/dbapi.py:247: in fetchone
    return next(self._iterator)
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/prestodb/client.py:467: in __iter__
    rows = self._query.fetch()
/opt/hostedtoolcache/Python/3.7.9/x64/lib/python3.7/site-packages/prestodb/client.py:534: in fetch
    status = self._request.process(response)
```

**Comment by inglesp** [comment]
What, is it 2007?

image

**Comment by inglesp** [comment]
Pulling the latest `prestosql/presto` means I can reproduce this locally.

Previously I had `5adc7a4d7326`.  I now have `52df0f7f5a6b`:

```
$ docker images prestosql/presto
REPOSITORY         TAG       IMAGE ID       CREATED        SIZE
prestosql/presto   latest    52df0f7f5a6b   2 days ago     1.06GB
prestosql/presto   <none>    5adc7a4d7326   8 months ago   1.03GB
```

Moreover, I can reproduce this on `master`, so the problem isn't with this PR.

**Comment by inglesp** [comment]
The [Python client library we're using](https://github.com/prestodb/presto-python-client/) hasn't been updated since July.

**Comment by inglesp** [comment]
After running the test setup (which creates some records in a MSSQL database), the following (which queries the MSSQL database via Presto) fails with the same error in `cursor.fetchall()`:

```
import prestodb

params = {
    "http_scheme": "http",
    "host": "localhost",
    "port": 8080,
    "catalog": "mssql",
    "schema": "dbo",
    "user": "ignored",
}

conn = prestodb.dbapi.connect(**params)
cursor = conn.cursor()

cursor.execute("select count(*) from patient_all_orgs_v2")

for row in cursor.fetchall():
    print(row)
```

**Comment by inglesp** [comment]
Distressingly, this script which creates records in MSSQL directly, does not reproduce the problem:

```
import pyodbc
import prestodb


mssql_conn = pyodbc.connect(
    "DRIVER={ODBC Driver 17 for SQL Server};"
    "SERVER=localhost,15785;"
    "DATABASE=Test_EMIS;"
    "UID=SA;"
    "PWD=Your_password123!"
)

try:
    mssql_cursor = mssql_conn.cursor()
    mssql_cursor.execute("create table t(c int)")
    mssql_cursor.execute("insert into t(c) values (1)")
    mssql_cursor.execute("insert into t(c) values (2)")
    mssql_cursor.execute("insert into t(c) values (3)")
    mssql_conn.commit()
except pyodbc.ProgrammingError as e:
    if not "There is already an object named 't' in the database" in str(e):
        raise


presto_conn = prestodb.dbapi.connect(
    http_scheme="http",
    host="localhost",
    port=8080,
    catalog="mssql",
    schema="dbo",
    user="ignored",
)
presto_cursor = presto_conn.cursor()
presto_cursor.execute("select count(*) from t")

for row in presto_cursor.fetchall():
    print(row)
```

**Comment by inglesp** [comment]
I've got a minimal repro.  In MSSQL, we create one table (`good`) with a column with no constraint, and a second table (`bad`) which is identical except there's a unique constraint on that column.  When querying the second table through Presto, we get the "Expected zero to one elements, but found multiple" error from Presto.

The script outputs:

```
$ python repro.py 
mssql good
[(0, )]
mssql bad
[(0, )]
presto good
[[0]]
presto bad
Traceback (most recent call last):
  File "repro2.py", line 49, in <module>
    print(presto_cursor.fetchall())
  File "/home/inglesp/.pyenv/versions/opensafely-cohort-extractor/lib/python3.8/site-packages/prestodb/dbapi.py", line 292, in fetchall
    return list(self.genall())
  File "/home/inglesp/.pyenv/versions/opensafely-cohort-extractor/lib/python3.8/site-packages/prestodb/client.py", line 467, in __iter__
    rows = self._query.fetch()
  File "/home/inglesp/.pyenv/versions/opensafely-cohort-extractor/lib/python3.8/site-packages/prestodb/client.py", line 534, in fetch
    status = self._request.process(response)
  File "/home/inglesp/.pyenv/versions/opensafely-cohort-extractor/lib/python3.8/site-packages/prestodb/client.py", line 412, in process
    raise self._process_error(response['error'], response.get('id'))
prestodb.exceptions.PrestoQueryError: PrestoQueryError(type=INTERNAL_ERROR, name=GENERIC_INTERNAL_ERROR, message="Expected zero to one elements, but found multiple", query_id=20201231_162207_00001_ae9jm)

```

This indicates that there isn't a problem with querying the MSSQL tables directly, but that the problem is with querying via Presto.

Here's the script:

```
import prestodb
import pyodbc

mssql_conn = pyodbc.connect(
    "DRIVER={ODBC Driver 17 for SQL Server};"
    "SERVER=localhost,15785;"
    "DATABASE=Test_EMIS;"
    "UID=SA;"
    "PWD=Your_password123!"
)

mssql_cursor = mssql_conn.cursor()
try:
    mssql_cursor.execute("drop table good")
except pyodbc.ProgrammingError:
    pass
try:
    mssql_cursor.execute("drop table bad")
except pyodbc.ProgrammingError:
    pass
mssql_cursor.execute("create table good (c integer null)")
mssql_cursor.execute("create table bad (c integer null, unique (c))")
mssql_conn.commit()

print("mssql good")
mssql_cursor.execute("select count(*) from good")
print(mssql_cursor.fetchall())

print("mssql bad")
mssql_cursor.execute("select count(*) from bad")
print(mssql_cursor.fetchall())

presto_conn = prestodb.dbapi.connect(
    http_scheme="http",
    host="localhost",
    port=8080,
    catalog="mssql",
    schema="dbo",
    user="ignored",
)
presto_cursor = presto_conn.cursor()

print("presto good")
presto_cursor.execute("select count(*) from good")
print(presto_cursor.fetchall())

print("presto bad")
presto_cursor.execute("select count(*) from bad")
print(presto_cursor.fetchall())
```

**Comment by inglesp** [comment]
The regression was introduced after release [348 (which works)](https://hub.docker.com/layers/prestosql/presto/348/images/sha256-052582da46fdde74e82994d67448cc080b6ef930ab54baec1ac97be81760563a) and [release 350](https://hub.docker.com/layers/prestosql/presto/350/images/sha256-adcf41a1fdf551be1d26a26b6ac75caae5d97fb2a6adc5151f5c9abb435704b4) (which doesn't).  Release 349 is not available.

**Comment by inglesp** [comment]
There is already an issue for this: https://github.com/trinodb/trino/issues/6464.

**Comment by inglesp** [comment]
The reason I didn't find this issue is because the original developers forked presto to make trino, and neither project seems to acknowledge the existence of the other.

For now, let's pin the docker image to the lastest presto version, and ask EMIS if there's an image we can use that matches the version they use.

**Comment by martint** [comment]
@inglesp, please see https://trino.io/blog/2020/12/27/announcing-trino.html for an explanation about why the name changed.

The docker images are maintained by what’s now the Trino project, and we’ll be releasing images under the new name soon.

**Comment by inglesp** [comment]
Thanks @martint.  Sounds an unpleasant position to be in.  Good luck with the rebranding!

File: cohortextractor/presto_utils.py [hunk 1]
```diff
@@ -1,19 +1,47 @@
...
+    if "PFX_PATH" in os.environ:
+        adapt_connection(conn, conn_params)
+
+    return ConnectionProxy(conn)
+
+
+def adapt_connection(conn, conn_params):
+    """Adapt connection to use passphrase-protected PKCS#12 certificate.
+
+    See instructions in opensafely/accessing-emis-data/ for getting a certificate.
```
Comment (ghickman) [inline cohortextractor/presto_utils.py:26]: Once the related manual bits are deployed, could we make this a full URL? Saves someone having to work out which docs/domain to look under!

File: cohortextractor/presto_utils.py [hunk 2]
```diff
@@ -1,19 +1,47 @@
...
+    """
+
+    with open(os.environ["PFX_PASSWORD_PATH"], "rb") as f:
+        pkcs12_password = f.read().strip()
+
+    session = requests.Session()
+    mount_prefix = "{http_scheme}://{host}:{port}".format(**conn_params)
```
Comment (ghickman) [inline cohortextractor/presto_utils.py:33]: Is the scheme dynamic here to support local use and testing?  (Or can it be fixed to one protocol?)

**Review by ghickman — APPROVED** [review]
👍

File: cohortextractor/presto_utils.py (hunk 1 above)
Comment (inglesp) [reply cohortextractor/presto_utils.py:26]: Good idea!

**Review by evansd — APPROVED** [review]

File: cohortextractor/presto_utils.py [hunk 3]
```diff
@@ -1,19 +1,47 @@
...
+    mount_adaptor = Pkcs12Adapter(
+        pkcs12_filename=os.environ["PFX_PATH"], pkcs12_password=pkcs12_password,
     )
+    session.mount(mount_prefix, mount_adaptor)
+    session.verify = False
+    conn._http_session = session
```
Comment (bloodearnest) [inline cohortextractor/presto_utils.py:39]: Ugh, I looked at the library code for this, unpleasant. Nothing we can do though

File: cohortextractor/presto_utils.py [hunk 4]
```diff
@@ -1,19 +1,47 @@
...
+    with open(os.environ["PFX_PASSWORD_PATH"], "rb") as f:
+        pkcs12_password = f.read().strip()
+
+    session = requests.Session()
+    mount_prefix = "{http_scheme}://{host}:{port}".format(**conn_params)
+    mount_adaptor = Pkcs12Adapter(
+        pkcs12_filename=os.environ["PFX_PATH"], pkcs12_password=pkcs12_password,
     )
+    session.mount(mount_prefix, mount_adaptor)
+    session.verify = False
```
Comment (bloodearnest) [inline cohortextractor/presto_utils.py:38]: Possibly more secure: `session.verify = '/path/to/server.crt'`

As opposed to accepting anything. But the risk here is low, so I leave it up to you.

**Review by bloodearnest — APPROVED** [review]
Good job, the presto/trino thing looks messy.

File: cohortextractor/presto_utils.py (hunk 2 above)
Comment (inglesp) [reply cohortextractor/presto_utils.py:33]: Yes -- we use http locally, and https in production.

File: cohortextractor/presto_utils.py (hunk 1 above)
Comment (inglesp) [reply cohortextractor/presto_utils.py:26]: Addressed in 92d8e6b.

File: cohortextractor/presto_utils.py (hunk 4 above)
Comment (inglesp) [reply cohortextractor/presto_utils.py:38]: See #392.

**Pull request merged**